p2p/peer_behaviour.go

```diff
@@ -0,0 +1,83 @@
+package p2p
+
+type ErrorBehaviourPeer int
+
+const (
+	ErrorBehaviourUnknown = iota
+	ErrorBehaviourBadMessage
+	ErrorBehaviourMessageOutofOrder
+)
+
+type GoodBehaviourPeer int
+
+const (
+	GoodBehaviourVote = iota
+	GoodBehaviourBlockPart
+)
+
+// PeerBehaviour provides an interface for reactors to signal the behaviour
+// of peers synchronously to other components.
+type PeerBehaviour interface {
+	Behaved(peer Peer, reason GoodBehaviourPeer)
+	Errored(peer Peer, reason ErrorBehaviourPeer)
+}
+
+type SwitchedPeerBehaviour struct {
+	sw *Switch
+}
+
+func (spb *SwitchedPeerBehaviour) Errored(peer Peer, reason ErrorBehaviourPeer) {
+	spb.sw.StopPeerForError(peer, reason)
+}
+
+func (spb *SwitchedPeerBehaviour) Behaved(peer Peer, reason GoodBehaviourPeer) {
+	spb.sw.MarkPeerAsGood(peer)
+}
+
+func NewSwitchedPeerBehaviour(sw *Switch) *SwitchedPeerBehaviour {
```

Comment:
should we return an interface here? then we'll be able to make the struct SwitchedPeerBehaviour private

Reply:
You can make it private and still return the concrete type. There is merrit to avoid return of interfaces.

Reply:
What are the benefits of avoiding to returning interfaces?

Reply:
If a caller needs the concrete type they can still acquire it without being forced to constrcut it manually in a potential inconsistent way. Additionally the implementation should not be couple to the interface which again might just be of interest to the caller.

Reply:
Thanks, that makes sense! If the type is private though, the caller can't do anything with the concrete type. Unless the caller is in the same package (using internals of the implementation in tests etc).

Reply:
Exactly right and inside the package boundary that might be okay.

Reply:
> What are the benefits of not returning interfaces (but private types instead)?

Being able to call methods not specified in the interface (private - within the package, public - everywhere). If this is not desired, which I think is the case here, returning an interface is preferable.

Reply:
For your consideration: http://idiomaticgo.com/post/best-practice/accept-interfaces-return-structs/

Reply:
I discussed this offline with @xla and pulled in @peterbourgon for another perspective to augment the article cited above. The most convincing argument I've heard is that interfaces are best used as consumer contracts, allowing the receiver to specify expectations, while return  types are producer specific concerns. Returning an interface couples the consumer expectation with the producing implementation. With this in mind I agree with @xla that we should return a private struct here.

Reply:
Or as @peterbourgon pointed out a public struct with private fields.

Reply:
Let's go with a public struct (with private fields). Returning a private struct doesn't seem uncommon but it still feels a bit off. For instance the documentation won't show in godoc. Sorry for the back and forth @brapse.

Reply:
@brapse and @xla asked why returning a private type was annoying. The primary reason is that it's difficult for callers to use, because the type can't escape the function boundary where it was received. That is, a function can't return a private type from another package, nor can it receive a private type from another package as a parameter.

Reply:
What a great discussion 👍 Thanks for your thoughts guys

> Returning an interface couples the consumer expectation with the producing implementation.

that's not always a bad thing (e.g. https://github.com/go-kit/kit/blob/22a2d43a4ffeb2460fdf0d332b4d4f0638253fb2/log/logfmt_logger.go#L37)

@peterbourgon nice to see you here 🧙‍♂️